### ISSUE TITLE
What are use cases for key import for repokey based repo?

### ISSUE THREAD

**Issue by griffrawk** [borgbackup/borg#2844]
I understand the use case for importing a missing or corrupted key for a keyfile based repo, but I fail to see in what circumstances import would be useful for for repokey scenario. There is other data in the config file besides the key, so if the config file were to be corrupted or lost an import is of little use.

In fact I tested by renaming config, and as expected the import failed.

Please forgive me if I'm missing the blindingly obvious.

**Comment by ThomasWaldmann** [comment]
maybe it currently can only cope with the (unlikely?) case that your key got removed from the config somehow. the other repo params are easy to reconstruct by looking at the dirs/files below data in a case of emergency (or, if you used the defaults, copy&paste + check from another repo's config).

but I agree, it could be more useful by doing that by itself when there is no config at all.

did you try whether key import works if you just removed the key from the config?

**Comment by griffrawk** [comment]
> did you try whether key import works if you just removed the key from the config?

Thanks. I just have. It imports the key if you just removed the key. 

On the basis that the id is also part of the keyfile and equally likely to be 'corrupted', if I remove the id as well, I get an exception when I import the keyfile in configparser.py (KeyError: 'id'), followed by another exception in archiver.py (configparser.NoOptionError: No option 'id' in section: 'repository'). I'd have hoped it restored the id as well.

(This is 1.0.9 from macports BTW)

I accept this is a bit of an artificial scenario!

**Comment by enkore** [comment]
The id isn't important, the rest of the config file, except the key, can be recovered manually (all settings are the default ~99 % of the time, and for the other 1 % you can guess). I suppose "borg check --repository-only" could be extended to deal with corrupted config files and guess settings, then a key can be imported ("--ignore-repository-id"?).

**Comment by rugk** [comment]
Would `--ignore-repository-id` also allow to import a key from one repo into another one?

Say, e.g., that you want to move your backup repo from one place to another. However, for some reason (e.g. bec. it is too much data to move and recreating it is faster), you decided to delete the old repo. You, however, still have (a backup) of the keyfile.
Now, you may still have printed your keyfile and put it far away in a safe place, so you want to avoid having to replace it by printing yet another one and replacing the old one.

So you maybe want to export the keyfile from the old repo and import it into a new one created with `borg init`?

**Comment by enkore** [comment]
No, that's not a good idea. It is the same case as copying an encrypted repository and updating both copies independently (just that one copy is not around any more); it's not secure.

A secure way to do this would be to `debug get-obj` the manifest from the old repo, import --iri, debug put-obj on the new repo.

**Comment by rugk** [comment]
Okay, but that's not really a thing a "usual" user would do. Also I don't really understand why that is secure (is not it essentially the same as copying the repo or, at least, the keyfile and some metadata.)

So maybe better don't allow the user to do this. :smiley:

**Comment by ThomasWaldmann** [comment]
As this is not specific to 1.0.x, I am moving this to 1.1.x milestone.

**Comment by ThomasWaldmann** [comment]
Answering the original question for borg2 (master):

`key import` on a repokey repo is useful in two cases:

1. The key blob in the repository was damaged or removed while the repository is otherwise intact — importing the exported key (keyfile or paperkey) restores access.
2. Moving a key between storage locations: `borg key import --key-location repository|keyfile` imports an exported key into either place, independent of where it was exported from.

The objection raised here in 2017 ("if the config file is lost, importing the key does not help") does not apply to borg2 anymore: the key is not kept in a config file. Keys live in the repository store's `keys/` namespace, matched by repository ID, and a repository can hold several keys with labels (`borg key list`).

Importing a key from a *different* repository (the `--ignore-repository-id` idea) is not planned — it has the same problems as independently updated copies of an encrypted repository, as discussed above.

Closing as answered.
